feat: Publish Docker image to Github Registry instead of Docker Hub

armenzg opened the pull request:
The overall changes in here are:

- Upload a [Docker image](https://github.com/orgs/getsentry/packages/container/package/action-release-image) to Github's registry
- Make this action use the Github registry image instead of the Docker Hub one

Fixes API-2791

armenzg commented on Dockerfile:
```diff
@@ -7,7 +7,7 @@ RUN export YARN_CACHE_FOLDER="$(mktemp -d)" \
 RUN yarn build
 
-FROM node:12-alpine
+FROM node:12-alpine as app
```
This may not be needed. I'm using `--target app` because I only want to tag this image and not the builder.

armenzg commented on .github/workflows/build.yml:
```diff
@@ -1,22 +1,47 @@
+          -t ghcr.io/getsentry/action-release-image:latest \
+          -t ghcr.io/getsentry/action-release-image:${{ github.sha }} \
+          --cache-from ghcr.io/getsentry/action-release-image:latest \
+          --target app
```
In a previous commit I was building both the builder and the app and publishing both, however, since the builder image is always being invalidated I will just build the app target.

asottile-sentry replied:
oh hmm, you could probably push both, then pull both, then use two `--cache-from` arguments to make it not invalidated each time?

armenzg commented on .github/workflows/build.yml:
```diff
@@ -3,20 +3,75 @@ on:
     branches:
       - master
+      # XXX: Remove this before merging
+      - armenzg/ci/docker-release-to-ghr
```
TODO: Remove after approval.

armenzg commented on .github/workflows/build.yml:
```diff
@@ -3,20 +3,75 @@ on:
+        # XXX: We need to test both master and v1 (official one)
+        # XXX: We need to create a v1 tag for the release to be completed
+        # XXX: Change this to master
+        uses: getsentry/action-release@armenzg/ci/docker-release-to-ghr
```
TODO: Remove before merging

armenzg commented on .github/workflows/build.yml:
```diff
@@ -3,20 +3,75 @@ on:
+
+      # Created "End to end action-release Integration" integration under Ecosystem's org
+      - name: Create Sentry release
+        # XXX: We need to test both master and v1 (official one)
```
We will do this on another PR.

armenzg commented on .github/workflows/test.yml:
```diff
@@ -4,14 +4,16 @@ on:
   push:
     branches:
       - master
-      - 'releases/*'
```
We have no branches named like this.

armenzg reviewed (commented):
Here's a created release from this branch:
<img width="1150" alt="image" src="https://user-images.githubusercontent.com/44410/179259001-7de64579-28ee-4d63-8cb6-4734f3aa74bb.png">

armenzg commented:
Anyone that can please review this? Thanks!

leeandher commented on .github/workflows/build.yml:
```diff
@@ -1,22 +1,76 @@
+      uses: actions/checkout@v3
       with:
+        # XXX: Why is this necessary?
         ref: ${{ github.head_ref }}
```
Found this ([docs](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context))
> The `head_ref`  or source branch of the pull request in a workflow run. This property  is only available when the event that triggers a workflow run is either pull_request or pull_request_target.

So it is the merge commit :+1:

asottile-sentry replied:
I would remove this and the line above -- github will by default check out the right thing

leeandher commented on .github/workflows/build.yml:
```diff
@@ -5,18 +5,70 @@ on:
+
+    - name: Login and publish
+      run: |
+        echo ${{ secrets.GITHUB_TOKEN }} | docker login ghcr.io -u $GITHUB_ACTOR --password-stdin
```
Is this `echo` statement leftover from debugging? Not sure it's meant to be here.

asottile-sentry replied:
it passes the password as stdin to `docker login` (which is expecting it there via `--password-stdin`) -- the idea there is to [avoid passwords in commands](https://youtu.be/SqXOchNKV_c)

leeandher commented on .github/workflows/build.yml:
```diff
@@ -5,18 +5,70 @@ on:
+
+  # End to end testing: what the users would do
+  # XXX: TODO: Notify the team if this fails
+  test-action-release:
```
Should this be it's own action so that we can test it before merging our PRs/community PRs as well?

armenzg replied:
Probably. Let me tackle this on a follow up PR

leeandher commented on .github/workflows/build.yml:
```diff
@@ -5,18 +5,70 @@ on:
 
-    - name: Docker
-      uses: docker/build-push-action@v1
+    # BUILDKIT_INLINE_CACHE creates the image in such a way that you can
+    # then use --cache-from (think of a remote cache)
+    # This feature can be used because of using the buildx plugin
+    # In the case of the Docker builder iamge, we should always see that it's using the remote cache
+    # NOTE: Sadly, the COPY line from the builder image (in the Dockerfile) always invalidates the cache,
+    # thus, we do not benefit from using --cache-from. The multi-stage build still gives us a smaller footprint
+    # TODO: Optimize the Dockerfile to take advantage of proper caching
```
Thanks for the explanation! It'll help a lot if anyone tries to debug/understand this later on 👏

asottile-sentry commented on .github/actions/use-local-dockerfile/action.yml:
```diff
@@ -1,13 +1,14 @@
+        ESC_DOCKER_IMAGE_LINK: 'docker\:\/\/ghcr.io\/getsentry\/action-release-image\:latest'
+      run: |
+        mv action.yml previous.yml
         sed "s/$ESC_DOCKER_IMAGE_LINK/Dockerfile/" previous.yml >> action.yml
```
fun little sed trick -- you can use any character as the delimiter so you don't need the complex escapes

so you can change the `IMAGE_LINK` to be the literal value and use `sed "s|$ESC_DOCKER_IMAGE_LINK|Dockerfile/"`

asottile-sentry replied:
this can also be `sed -i` and then you can remove the `mv` commands (note this requires gnu sed so won't work on macos, but that's fine here!)

leeandher reviewed (commented):
Left a few comments/questions, only concern is the stray `echo` command. Looks good!

asottile-sentry commented on .github/workflows/build.yml:
```diff
@@ -5,18 +5,70 @@ on:
+
+    - name: Login and publish
+      run: |
+        echo ${{ secrets.GITHUB_TOKEN }} | docker login ghcr.io -u $GITHUB_ACTOR --password-stdin
```
neat little shell trick -- `docker login ... <<< ${{ secrets.GITHUB_TOKEN }}` -- https://youtu.be/Xf_82stIbB8

asottile-sentry commented on Dockerfile:
```diff
@@ -7,7 +7,7 @@ RUN export YARN_CACHE_FOLDER="$(mktemp -d)" \
 COPY . .
```
ah -- probably need a `.dockerignore` with `.git` in it? -- https://youtu.be/dEZ5WVWLris

asottile-sentry reviewed (approved):
a bunch of small things -- but looks good overall!

![](https://i.fluffy.cc/zVmg9csn3pdFq4S6hQKHKzdCl3B7cRvG.gif)

asottile-sentry commented on .dockerignore:
```diff
@@ -0,0 +1,5 @@
+# In other words, if a file should not be inside of the Docker container it should be
+# added to the list, otherwise, it will invalidate cache layers and have to rebuild
+# all layers after a COPY command
+.github
```
ah the `.git` directory is the more important one here

asottile-sentry commented on .github/workflows/build.yml:
```diff
@@ -51,8 +51,11 @@ jobs:
+        docker push ghcr.io/getsentry/action-release-image:latest || rm .docker/config.json
+        # WARNING! Your password will be stored unencrypted in /home/runner/.docker/config.json.
+        # Read https://docs.docker.com/engine/reference/commandline/login/#credentials-store
+        rm .docker/config.json
```
this will need to be `~/.docker/config.json`

armenzg reviewed (commented):
I've removed the job that would test the action from master.

Something got messed up when it started trying to report the release from the pull_request.
I will look into it in the future.

armenzg commented:
@leeandher @asottile-sentry I've addressed your concerns. I've reduced the scope by not testing the action and I've started publishing the builder Docker image. `--cache-from` is now working properly thanks to the `.dockerignore` change.

asottile-sentry reviewed (approved):
![](https://i.fluffy.cc/3X84Bq2RR182XcPr0NTKKCp7wvNT8MhD.gif)
